exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/Queries.kt

```diff
@@ -541,18 +602,37 @@ fun <T : Table, E : Any> T.batchUpsert(
 /**
  * Represents the SQL statement that either batch inserts new rows into a table, or updates the existing rows if insertions violate unique constraints.
  *
+ * To set specific columns with values on update that differ from the values that would be inserted, use `UpsertStatement.onUpdate()`
+ * within the [body] lambda block. If `onUpdate()` is omitted, all columns will be updated with the values provided for the insert.
+ * To specify manually that the insert value should be used when setting an update, for example in a function or expression,
+ * invoke `insertValue()` with the desired column as the function argument.
+ *
  * @param data Sequence of values to use in batch upsert.
  * @param keys (optional) Columns to include in the condition that determines a unique constraint match. If no columns are provided,
  * primary keys will be used. If the table does not have any primary keys, the first unique index will be attempted.
- * @param onUpdate List of pairs of specific columns to update and the expressions to update them with.
- * If left null, all columns will be updated with the values provided for the insert.
  * @param onUpdateExclude List of specific columns to exclude from updating.
  * If left null, all columns will be updated with the values provided for the insert.
  * @param where Condition that determines which rows to update, if a unique violation is found.
  * @param shouldReturnGeneratedValues Specifies whether newly generated values (for example, auto-incremented IDs) should be returned.
  * See [Batch Insert](https://github.com/JetBrains/Exposed/wiki/DSL#batch-insert) for more details.
  * @sample org.jetbrains.exposed.sql.tests.shared.dml.UpsertTests.testBatchUpsertWithSequence
  */
+fun <T : Table, E : Any> T.batchUpsert(
+    data: Sequence<E>,
+    vararg keys: Column<*>,
+    onUpdateExclude: List<Column<*>>? = null,
+    where: (SqlExpressionBuilder.() -> Op<Boolean>)? = null,
+    shouldReturnGeneratedValues: Boolean = true,
+    body: BatchUpsertStatement.(E) -> Unit
+): List<ResultRow> {
+    return batchUpsert(data.iterator(), null, onUpdateExclude, where, shouldReturnGeneratedValues, keys = keys, body = body)
+}
+
+@Deprecated(
+    "This `batchUpsert()` with `onUpdate` parameter will be removed in future releases. " +
+        "Please use function `onUpdate()` in `body` lambda block instead.",
+    level = DeprecationLevel.WARNING
+)
```

Comment:
I'd like to use `ReplaceWith` but this may be too complex a change. The closest I could get is with the following:
```kt
ReplaceWith("upsertReturning(keys = keys, onUpdateExclude = onUpdateExclude, where = where) { body.invoke();it.onUpdate { onUpdate.invoke() } }")

// takes this original
Items.upsert(
    onUpdate = { listOf(Items.price to Items.price.times(10.0)) }
) {
    it[id] = 1
    it[name] = "B"
    it[price] = 200.0
}

// and converts it to this
Items.upsert(
    keys = arrayOf()
) {
    it[id] = 1
    it[name] = "B"
    it[price] = 200.0
    it.onUpdate {
        listOf(price to price.times(10.0)) // getting this to convert is difficult
    }
}
```